Switch the Supervisor/task process from line-based to length-prefixed

ashb opened the pull request:
Relates to #46426

The existing JSON Lines based approach had three major drawbacks

1. In the case of really large lines (in the region of 10 or 20MB) the python
   line buffering could _sometimes_ result in a partial read
2. The JSON based approach didn't have the ability to add any metadata (such
   as errors).
3. Not every message type/call-site waited for a response, which meant those
   client functions could never get told about an error

One of the ways this line-based approach fell down was if you suddenly tried
to run 100s of triggers at the same time you would get an error like this:

```
Traceback (most recent call last):
  File "/Users/ash/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/streams.py", line 568, in readline
    line = await self.readuntil(sep)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ash/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/streams.py", line 663, in readuntil
    raise exceptions.LimitOverrunError(
asyncio.exceptions.LimitOverrunError: Separator is found, but chunk is longer than limit
```

The other way this caused problems was if you parse a large dag (as in one
with 20k tasks or more) the DagFileProcessor could end up getting a partial
read which would be invalid JSON.

This changes the communications protocol in in a couple of ways.

First off at the python level the separate send and receive methods in the
client/task side have been removed and replaced with a single `send()` that
sends the request, reads the response and raises an error if one is returned.
(But note, right now almost nothing in the supervisor side sets the error,
that will be a future PR.)

Secondly the JSON Lines approach has been changed from a line-based protocol
to a binary "frame" one. The protocol (which is the same for whichever side is
sending) is length-prefixed, i.e. we first send the length of the data as a
4byte big-endian integer, followed by the data itself. This should remove the
possibility of JSON parse errors due to reading incomplete lines

Finally the last change made in this PR is to remove the "extra" requests
socket/channel. Upon closer examination with this comms path I realised that
this socket is unnecessary: Since we are in 100% control of the client side we
can make use of the bi-directional nature of `socketpair` and save file
handles. This also happens to help the `run_as_user` feature which is
currently broken, as without extra config to `sudoers` file, `sudo` will close
all filehandles other than stdin, stdout, and stderr -- so by introducing this
change we make it easier to re-add run_as_user support.

In order to support this in the DagFileProcessor (as the fact that the proc
manager uses a single selector for multiple processes) means I have moved the
`on_close` callback to be part of the object we store in the `selector` object
in the supervisors, previoulsy it was the "on_read" callback, now we store a
tuple of `(on_read, on_close)` and on_close is called once universally.

This also changes the way comms are handled from the (async) TriggerRunner
process. Previously we had a sync+async lock, but that made it possible to end
up deadlocking things. The change now is to have `send` on
`TriggerCommsDecoder` "go back" to the async even loop via `async_to_sync`, so
that only async code deals with the socket, and we can use an async lock
(rather than the hybrid sync and async lock we tried before). This seems to
help the deadlock issue, but I'm not 100% sure it will remove it entirely, but
it makes it much much harder to hit - I've not been able to reprouce it with
this change.

Fixes #50185, fixes #51213, closes #51279

ashb committed:
Switch the Supervisor/task process from line-based to length-prefixed

The existing JSON Lines based approach had two major drawbacks

1. In the case of really large lines (in the region of 10 or 20MB) the python
   line buffering could _sometimes_ result in a partial read
2. The JSON based approach didn't have the ability to add any metadata (such
   as errors).
3. Not every message type/call-site waited for a response, which meant those
   client functions could never get told about an error

One of the ways this line-based approach fell down was if you suddenly tried
to run 100s of triggers at the same time you would get an error like this:

```
Traceback (most recent call last):
  File "/Users/ash/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/streams.py", line 568, in readline
    line = await self.readuntil(sep)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ash/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/streams.py", line 663, in readuntil
    raise exceptions.LimitOverrunError(
asyncio.exceptions.LimitOverrunError: Separator is found, but chunk is longer than limit
```

The other way this caused problems was if you parse a large dag (as in one
with 20k tasks or more) the DagFileProcessor could end up getting a partial
read which would be invalid JSON.

This changes the communications protocol in in a couple of ways.

First off at the python level the separate send and receive methods in the
client/task side have been removed and replaced with a single `send()` that
sends the request, reads the response and raises an error if one is returned.
(But note, right now almost nothing in the supervisor side sets the error,
that will be a future PR.)

Secondly the JSON Lines approach has been changed from a line-based protocol
to a binary "frame" one. The protocol (which is the same for whichever side is
sending) is length-prefixed, i.e. we first send the length of the data as a
4byte big-endian integer, followed by the data itself. This should remove the
possibility of JSON parse errors due to reading incomplete lines

Finally the last change made in this PR is to remove the "extra" requests
socket/channel. Upon closer examination with this comms path I realised that
this socket is unnecessary: Since we are in 100% control of the client side we
can make use of the bi-directional nature of `socketpair` and save file
handles. This also happens to help the `run_as_user` feature which is
currently broken, as without extra config to `sudoers` file, `sudo` will close
all filehandles other than stdin, stdout, and stderr -- so by introducing this
change we make it easier to re-add run_as_user support.

In order to support this in the DagFileProcessor (as the fact that the proc
manager uses a single selector for multiple processes) means I have moved the
`on_close` callback to be part of the object we store in the `selector` object
in the supervisors, previoulsy it was the "on_read" callback, now we store a
tuple of `(on_read, on_close)` and on_close is called once universally.

This also changes the way comms are handled from the (async) TriggerRunner
process. Previously we had a sync+async lock, but that made it possible to end
up deadlocking things. The change now is to have `send` on
`TriggerCommsDecoder` "go back" to the async even loop via `async_to_sync`, so
that only async code deals with the socket, and we can use an async lock
(rather than the hybrid sync and async lock we tried before). This seems to
help the deadlock issue, but I'm not 100% sure it will remove it entirely, but
it makes it much much harder to hit - I've not been able to reprouce it with
this change

ashb commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -1272,7 +1291,6 @@ class InProcessTestSupervisor(ActivitySubprocess):
     """A supervisor that runs tasks in-process for easier testing."""
 
     comms: InProcessSupervisorComms = attrs.field(init=False)
-    stdin = attrs.field(init=False)
```
@kaxil I assume `dag.test()` is covered by unit tests right? I'm 80% sure this change is safe. Ditto for the removal of lock on L`1246`

ashb replied:
They are, and this fails without this

ashb commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -946,16 +964,11 @@ def _monitor_subprocess(self):
                     self._process_exit_monotonic
                     and time.monotonic() - self._process_exit_monotonic > SOCKET_CLEANUP_TIMEOUT
                 ):
-                    log.debug(
-                        "Forcefully closing remaining sockets",
-                        open_sockets=self._num_open_sockets,
-                        pid=self.pid,
-                    )
```
This debug log added nothing -- we'd get a log message at warning from within cleanup_open_sockets if it did anything

ashb commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -523,24 +532,26 @@ def _register_pipe_readers(self, stdout: socket, stderr: socket, requests: socke
-            logs.fileno(): "logs",
-        }
+        # Track the open sockets, and for debugging what type each one is
+        self._open_sockets.update(
```
`update` because this is a WeakKeyDict -- meaning the dict won't keep things open/alive if the socket object gets GCd.

ashb commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -224,14 +229,13 @@ def _configure_logs_over_json_channel(log_fd: int):
         ("stdout", 1, child_stdout, "w"),
         ("stderr", 2, child_stderr, "w"),
     ):
-        handle = getattr(sys, handle_name)
-        handle.close()
```
`dup2` handles the closing for us, and does a better job of dealing with race conditions too.

kaxil commented on airflow-core/src/airflow/dag_processing/processor.py:
```diff
@@ -102,18 +101,16 @@ class DagFileParsingResult(BaseModel):
     )
 
-    msg = comms_decoder.get_message()
+    msg = comms_decoder._get_response()
```
Why not  `msg = comms_decoder.send()`

ashb replied:
Because for this very first message (and only in this case) the parent/supervisor sends this information unsolicited

kaxil replied:
Worth adding a code comment, so it doesn't look weird here on why we access an internal method

kaxil commented on airflow-core/src/airflow/dag_processing/processor.py:
```diff
@@ -301,18 +296,24 @@ def _handle_request(self, msg: ToManager, log: FilteringBoundLogger) -> None:  #
             log.error("Unhandled request", msg=msg)
+            self.send_msg(
+                None,
+                in_response_to=req_id,
```
Should we call this arg as `request_id` instead of `in_response_to`?

ashb replied:
Could do, yes.

kaxil commented on airflow-core/src/airflow/dag_processing/processor.py:
```diff
@@ -266,20 +263,18 @@ def _on_child_started(
             callback_requests=callbacks,
         )
-        self.send_msg(msg)
+        self.send_msg(msg, in_response_to=0)
```
Is `in_response_to` auto-incrementing or we can provide any request_id like `0` here?

kaxil replied:
I am reviewing files in GH's diff order -- so if I later find the answer for this, will add a comment

kaxil replied:
what happens if those requests are sent with same ids

ashb replied:
It means nothing to the server, so it's up to the client to ensure uniqueness

kaxil commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -691,14 +690,58 @@ class TriggerDetails(TypedDict):
+        bytes = frame.as_bytes()
+
+        async with self._lock:
+            self._async_writer.write(bytes)
```
I think we need `await self._async_writer.drain()` after this line too based on https://docs.python.org/3/library/asyncio-stream.html#asyncio.StreamWriter.write

ashb replied:
I don't we _need_ it. The drain will happen in the background, and we can just as well wait on reading as block on the buffer being emptied and then waiting on reading.

kaxil commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -691,14 +690,58 @@ class TriggerDetails(TypedDict):
+        len = int.from_bytes(len_bytes, byteorder="big")
+
+        buffer = await self._async_reader.readexactly(len)
+        return self.resp_decoder.decode(buffer)
```
Can this hang in any scenario?

ashb replied:
decode? No, or at least it's not doing any IO

kaxil replied:
No no, that block I mean:

```py
        len_bytes = await self._async_reader.readexactly(4)
        len = int.from_bytes(len_bytes, byteorder="big")

        buffer = await self._async_reader.readexactly(len)
```

ashb replied:
It's async, so it'll "hang" in the sense that it will block the current task until the response is availabel, but it's async so other triggers will continue to run.

I still don't think I've understood your question though

kaxil commented on airflow-core/src/airflow/jobs/triggerer_job_runner.py:
```diff
@@ -691,14 +690,58 @@ class TriggerDetails(TypedDict):
+        len_bytes = await self._async_reader.readexactly(4)
+        len = int.from_bytes(len_bytes, byteorder="big")
+
+        buffer = await self._async_reader.readexactly(len)
```
Should we set an upper limit for the size?

kaxil commented on airflow-core/tests/unit/dag_processing/test_manager.py:
```diff
@@ -577,44 +575,40 @@ def test_kill_timed_out_processors_no_kill(self):
-
-        assert val == expected_buffer
+        # Read response from the read end of the socket
+        read_socket.settimeout(0.1)
```
It is already set on L604

kaxil commented on devel-common/src/tests_common/pytest_plugin.py:
```diff
@@ -1956,17 +1956,20 @@ def override_caplog(request):
+    from airflow.sdk.execution_time.comms import CommsDecoder
+
+    comms = mock.create_autospec(CommsDecoder)
+    monkeypatch.setattr(airflow.sdk.execution_time.task_runner, "SUPERVISOR_COMMS", comms, raising=False)
```
```suggestion
    from airflow.sdk.execution_time import task_runner
    from airflow.sdk.execution_time.comms import CommsDecoder

    comms = mock.create_autospec(CommsDecoder)
    monkeypatch.setattr(task_runner, "SUPERVISOR_COMMS", comms, raising=False)
```

kaxil replied:
nit -- feel free to ignore

ashb replied:
Probably worth it. If I'm making any other changes/fixes to this PR I'll do it

kaxil commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -80,20 +90,152 @@
+        # convert the complex to a tuple of real, imag
+        return datetime(
+            obj.year, obj.month, obj.day, obj.hour, obj.minute, obj.second, obj.microsecond, tzinfo=obj.tzinfo
+        )
```
Is this because we were previously relying purely on Pydantic for ser/de-ser? 

Everything should be part of BaseModel though? When would I need (for example) to add a new type here?

ashb replied:
This is because we now call `msg.model_dump()` instead of `msg.model_dump_json()`, so we now get a dict with Datetime objects etc. We need to add something if we add a new message type and it fails to send -- it'll be pretty obvious during dev.

I notice the comment on L104 is wrong (i copied it verbatim from the msgspec docs)

kaxil commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -80,20 +90,152 @@
+        self.req_encoder.encode_into(self, buffer, 4)
+
+        n = len(buffer) - 4
+        if n > 2**32:
```
Shouldn't this be `>=`?

ashb replied:
Probably, yes

kaxil commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -80,20 +90,152 @@
+        return buffer
+
+
+class _ResponseFrame(_RequestFrame, msgspec.Struct, array_like=True, frozen=True, omit_defaults=True):
```
```suggestion
class _ResponseFrame(_RequestFrame):
```

Should be ok?

ashb replied:
Might be. I'll see.

kaxil commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -80,20 +90,152 @@
+    """Handle communication between the task in this process and the supervisor parent process."""
+
+    log: Logger = attrs.field(repr=False, factory=structlog.get_logger)
+    request_socket: socket = attrs.field(factory=lambda: socket(fileno=0))
```
This is now both: request and response. I don't have a good suggestion for the name (`comms_socket` -- maybe?) -- so probably fine to keep it🤷

kaxil commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -80,20 +90,152 @@
+                f"unable to read full response in child. (We read {nread}, but expected {len})"
+            )
+        if nread == 0:
+            raise EOFError("Request socket closed before response was complete")
```
Probably worth logging the id?

kaxil commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -346,10 +331,12 @@ def _fork_main(
     # Store original stderr for last-chance exception handling
     last_chance_stderr = _get_last_chance_stderr()
 
+    # os.environ["_AIRFLOW_SUPERVISOR_FD"] = str(requests.fileno())
```
?

ashb replied:
Oh yeah, was playing around with other ways of doing things. shouldn't have committed this

kaxil commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -1441,7 +1445,6 @@ def cb(sock: socket):
                 with suppress(StopIteration):
                     gen.send(buffer)
             # Tell loop to close this selector
```
```suggestion
```

kaxil commented on task-sdk/src/airflow/sdk/execution_time/task_runner.py:
```diff
@@ -725,31 +649,33 @@ def send_request(self, log: Logger, msg: SendMsgType):
 
 def startup() -> tuple[RuntimeTaskInstance, Context, Logger]:
-    msg = SUPERVISOR_COMMS.get_message()
+    # The parent sends us a StartupDetails message un-prompted. After this, ever single message is only sent
```
```suggestion
    # The parent sends us a StartupDetails message un-prompted. After this, every single message is only sent
```

kaxil commented on task-sdk/tests/task_sdk/execution_time/test_supervisor.py:
```diff
@@ -1729,33 +1763,23 @@ def test_handle_requests(
-
-        if val:
+        if frame.body is not None:
             # Using BytesIO to simulate a readable stream for CommsDecoder.
```
```suggestion
```

kaxil reviewed (approved):
🎉

ashb committed:
Deal with compat in tests

This compat issue is only in tests, as nothing in the runtime of airflow-core
imports/calls methods directly on SUPERVISOR_COMMS, we are only importing it
in tests to mkae assertions about the behavour/to stub the return values.

ashb committed:
Code review

amoghrajesh commented on task-sdk/src/airflow/sdk/execution_time/comms.py:
```diff
@@ -19,12 +19,17 @@
   be running user's code, so we can't read from stdin until we enter our code, such as when requesting an XCom
   value etc.)
+* Every request returns a response, even if the frame is otherwise empty.
+* Requests are written by the subprocess to fd0/stdin. This is making use of the fact that stdin is a
+  bi-directional socket, and thus we can write to it and don't need a dedicated extra socket for sending
+  requests.
```
Nice, love it

uranusjr commented on airflow-core/src/airflow/dag_processing/processor.py:
```diff
@@ -266,20 +263,18 @@ def _on_child_started(
         dump_opts = {}
         if isinstance(msg, DagFileParsingResult):
             self.parsing_result = msg
-            return
-        if isinstance(msg, GetConnection):
+        elif isinstance(msg, GetConnection):
```
Is this change correct? It’s going to send `None` as a request if `msg` is a DagFileParsingResult.

amoghrajesh replied:
Good qn, i wonder it it will raise an error here: https://github.com/apache/airflow/pull/51699/files#diff-c076ff9d88eb776c2f874ea4fbb66c9d090e080ed5c9b80c2c36dcb3a51360d3R228-R230

ashb replied:
It is correct. Part of the change in this is that _every_ message must send a response, even if that response is empty. (It's wrapped up in a `_ResponseFrame`, where body can be None by`self.send_msg`. But we must send something)

ashb replied:
This was a case where previously we didn't have any response

amoghrajesh commented on task-sdk/tests/task_sdk/execution_time/test_comms.py:
```diff
@@ -0,0 +1,83 @@
+            "dag_rel_path": "/dev/null",
+            "bundle_info": {"name": "any-name", "version": "any-version"},
+        }
+        bytes = msgspec.msgpack.encode(_ResponseFrame(0, msg, None))
+        w.sendall(len(bytes).to_bytes(4, byteorder="big") + bytes)
+
+        decoder = CommsDecoder(socket=r, log=None)
+
+        msg = decoder._get_response()
+        assert isinstance(msg, StartupDetails)
+        assert msg.ti.id == uuid.UUID("4d828a62-a417-4936-a7a6-2b3fabacecab")
+        assert msg.ti.task_id == "a"
+        assert msg.ti.dag_id == "c"
+        assert msg.dag_rel_path == "/dev/null"
+        assert msg.bundle_info == BundleInfo(name="any-name", version="any-version")
+        assert msg.start_date == timezone.datetime(2024, 12, 1, 1)
```
Nice, gives confidence!

amoghrajesh commented on devel-common/src/tests_common/pytest_plugin.py:
```diff
@@ -1956,17 +1956,27 @@ def override_caplog(request):
+        comms = mock.create_autospec(CommsDecoder)
+        comms.send = comms.get_response
+        monkeypatch.setattr(task_runner, "SUPERVISOR_COMMS", comms, raising=False)
+    yield comms
```
☠️ it is what it is!

amoghrajesh reviewed (approved):
LGTM +1.

ashb committed:
fixup! Deal with compat in tests

gopidesupavan commented on task-sdk/src/airflow/sdk/execution_time/supervisor.py:
```diff
@@ -552,37 +542,52 @@ def _register_pipe_readers(self, stdout: socket, stderr: socket, requests: socke
             except Exception:
-                log.exception("Unable to decode message", line=line)
+                log.exception("Unable to decode message", body=request.body)
                 continue
```
Why we are continuing? i think if we unable to decode we should respond ?
